common/djangoapps/third_party_auth/saml.py

```diff
@@ -93,3 +95,132 @@ def wrapped_method(*args, **kwargs):
+                missing
+            )
+            return missing
+
+    def get_odata_api_client(self, user_id='sfadmin'):
+        """
+        Get a Requests session with the headers needed to properly authenticate it with
+        the SAP SuccessFactors OData API.
+        """
+        assertion = requests.post(
+            self.sapsf_idp_url,
+            data={
+                'client_id': self.odata_client_id,
+                'user_id': user_id,
+                'token_url': self.sapsf_token_url,
+                'private_key': self.sapsf_private_key,
+            }
+        ).text
+        token = requests.post(
+            self.sapsf_token_url,
+            data={
+                'client_id': self.odata_client_id,
+                'company_id': self.odata_company_id,
+                'grant_type': 'urn:ietf:params:oauth:grant-type:saml2-bearer',
+                'assertion': assertion,
+            }
+        ).json()['access_token']
+        session = requests.Session()
+        session.headers.update({'Authorization': 'Bearer {}'.format(token), 'Accept': 'application/json'})
+        return session
+
+
+    def get_user_details(self, attributes):
+        """
+        Attempt to get rich user details from the SAP SuccessFactors OData API. If we're missing any
+        of the details we need to do that, fail nicely by returning the details we're able to extract
+        from just the SAML response and log a warning.
+        """
+        details = super(SapSuccessFactorsIdentityProvider, self).get_user_details(attributes)
+        if self.missing_variables():
+            return details
+        username = details['username']
+        client = self.get_odata_api_client(user_id=username)
+        response = client.get(
+            '{root_url}User(userId=\'{user_id}\')?$select=username,firstName,lastName,defaultFullName,email'.format(
+                root_url=self.odata_api_root_url,
+                user_id=username
+            )
+        ).json()
+        return {
+            'username': response['d']['username'],
+            'first_name': response['d']['firstName'],
+            'last_name': response['d']['lastName'],
+            'fullname': response['d']['defaultFullName'],
+            'email': response['d']['email'],
+        }
+
+def get_saml_idp_choices():
+    """
+    Get a list of the available SAMLIdentityProvider subclasses that can be used to process
+    SAML requests, for use in the Django administration form.
+    """
+    return (
+        (STANDARD_SAML_PROVIDER_KEY, 'Standard SAML provider'),
+        ('sap_success_factors', 'SAP SuccessFactors provider'),
+    )
+
+def get_saml_idp_class(idp_identifier_string):
+    """
+    Given a string ID indicating the type of identity provider in use during a given request, return
+    the SAMLIdentityProvider subclass able to handle requests for that type of identity provider.
+    """
+    choices = {
```

Comment:
Probably. I'm thinking error-level; we should only be calling this method from one place, and that place should only be calling with values that are valid options.